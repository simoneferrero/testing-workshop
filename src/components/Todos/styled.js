import styled from 'styled-components/macro'

import { SPACING } from 'constants/index'

export const StyledWrapper = styled.div`
	display: grid;
	grid-gap: ${SPACING.base}px;
	grid-template-columns: 1fr;
	justify-items: center;
	margin: 0 auto;
	margin-top: ${SPACING.headerHeight}px;
	padding: ${SPACING.base}px;
	max-width: 500px;
`

export const StyledLabel = styled.label`
	font-size: 12px;
`

export const StyledPaper = styled.div`
	background-color: #fff;
	border-radius: ${SPACING.half}px;
	border-top-right-radius: 0px;
	border-bottom-left-radius: 0px;
	color: #000;
	padding: ${SPACING.half}px 0px;
	width: 100%;
`
