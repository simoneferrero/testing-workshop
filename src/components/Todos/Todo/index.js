import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components/macro'

import { SPACING } from 'constants/index'

const StyledTodo = styled.button`
	background-color: inherit;
	border: none;
	color: ${({ completed }) => completed && '#808080'};
	cursor: pointer;
	font-family: 'Nanum Pen Script', cursive;
	font-size: 24px;
	margin-bottom: ${SPACING.half}px;
	padding: ${SPACING.half}px ${SPACING.base}px;
	text-align: left;
	text-decoration: ${({ completed }) => completed && 'line-through'};
	transition: box-shadow 0.3s ease-in-out, color 0.3s ease-in-out,
		transform 0.3s ease-in-out;
	width: 100%;
	will-change: transform;

	:hover {
		box-shadow: 0px 4px 6px rgba(0, 0, 0, 0.3);
		transform: translate(5px, -5px) scale(1.01);
	}
`

const Todo = ({ completed, handleClick, id, text }) => (
	<StyledTodo completed={completed} onClick={() => handleClick(id, !completed)}>
		{text}
	</StyledTodo>
)

Todo.propTypes = {
	completed: PropTypes.bool.isRequired,
	handleClick: PropTypes.func.isRequired,
	id: PropTypes.string.isRequired,
	text: PropTypes.string.isRequired,
}

export default Todo
