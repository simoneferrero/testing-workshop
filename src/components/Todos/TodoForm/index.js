import React, { useState } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components';

import { COLORS, SPACING } from '../../../constants/index'

const StyledForm = styled.form`
	display: flex;
	width: 100%;
`

const StyledInput = styled.input`
	border-radius: ${SPACING.half}px 0px 0px ${SPACING.half}px;
	border: none;
	box-sizing: border-box;
	height: 40px;
	padding: ${SPACING.base}px;
	width: 100%;
`

const StyledButton = styled.button`
	background-color: ${COLORS.accent};
	border-radius: 0px 8px 8px 0px;
	border: 0px;
	box-sizing: border-box;
	color: ${COLORS.accentText};
	cursor: pointer;
	height: 40px;
	padding: ${SPACING.base};
`

const TodoForm = ({ handleSubmit }) => {
	const [newTodo, setNewTodo] = useState('')
	const onSubmit = (event) => {
		event.preventDefault()

		if (!!newTodo) {
			handleSubmit(newTodo)
			setNewTodo('')
		}
	}

	return (
		<StyledForm onSubmit={ onSubmit }>
			<StyledInput
				onChange={ ({ target: { value } }) => setNewTodo(value) }
				placeholder="Add TODO"
				value={ newTodo }
			/>
			<StyledButton type="submit">ADD</StyledButton>
		</StyledForm>
	)
}

TodoForm.propTypes = {
	handleSubmit: PropTypes.func.isRequired,
}

export default TodoForm
