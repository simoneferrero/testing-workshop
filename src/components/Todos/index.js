import React from 'react'

import Todo from './Todo'
import TodoForm from './TodoForm'
import { StyledLabel, StyledPaper, StyledWrapper } from './styled'

class Todos extends React.Component {
	constructor() {
		super()

		this.handleClick = this.handleClick.bind(this)
		this.handleSubmit = this.handleSubmit.bind(this)
	}

	state = {
		showCompleted: true,
		todos: [],
	}

	async handleClick(id, completed) {
		this.setState(({ todos }) => ({
			todos: todos.map((todo) =>
				todo.id === id ? { ...todo, completed } : todo
			),
		}))
	}

	async handleSubmit(text) {
		const todo = {
			id: String(Date.now()),
			text,
			completed: false,
		}
		this.setState(({ todos }) => ({ todos: [...todos, todo] }))
	}

	render() {
		const { showCompleted, todos } = this.state

		return (
			<StyledWrapper data-testid="todos">
				<TodoForm handleSubmit={this.handleSubmit} />
				<StyledLabel htmlFor="showCompleted">
					Show Completed:{' '}
					<input
						checked={showCompleted}
						id="showCompleted"
						onChange={() => this.setState({ showCompleted: !showCompleted })}
						type="checkbox"
					/>
				</StyledLabel>
				<StyledPaper data-testid="paper">
					{todos.map((todo) => (
						<Todo handleClick={this.handleClick} key={todo.id} {...todo} />
					))}
				</StyledPaper>
			</StyledWrapper>
		)
	}
}

export default Todos
