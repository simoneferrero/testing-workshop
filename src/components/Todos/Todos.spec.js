import { within } from '@testing-library/react'

import Todos from './index'

describe('Given <Todos />', () => {
	const renderComponent = () => {
		const renderedComponent = render(<Todos />)
		const {
			getAllByTestId,
			getByLabelText,
			getByPlaceholderText,
			getByTestId,
			getByText,
		} = renderedComponent

		return {
			...renderedComponent,
			getAddTodoButton: () => getByText(/ADD/i),
			getAddTodoInput: () => getByPlaceholderText(/Add TODO/i),
			getAllTodos: () => getAllByTestId(/(todo-)/i),
			getShowCompletedCheckbox: () => getByLabelText(/(Show Completed:)/i),
			getTodo: (text) => within(getByTestId('paper')).getByText(text),
			getWrapper: () => getByTestId('todos'),
		}
	}
	it('should render correctly', () => {
		const {
			container,
			getAddTodoButton,
			getAddTodoInput,
			getShowCompletedCheckbox,
			getWrapper,
		} = renderComponent()

		expect(container.firstChild).toMatchSnapshot()
		expect(getWrapper()).toBeInTheDocument()
		expect(getAddTodoInput()).toBeInTheDocument()
		expect(getAddTodoButton()).toBeInTheDocument()
		expect(getShowCompletedCheckbox()).toBeInTheDocument()
	})

	it('should correctly add new todos and mark them complete', () => {
		const todos = ['First todo', 'Second todo']
		const {
			getAllTodos,
			getAddTodoButton,
			getAddTodoInput,
			getTodo,
		} = renderComponent()

		todos.forEach((todo) => {
			userEvent.type(getAddTodoInput(), todo)
			userEvent.click(getAddTodoButton())

			const renderedTodo = getTodo(todo)

			expect(renderedTodo).toBeInTheDocument()
			expect(renderedTodo).not.toHaveStyle('color: #808080;')
			expect(renderedTodo).not.toHaveStyle('text-decoration: line-through;')
		})

		expect(getAllTodos().length).toBe(todos.length)

		const completedTodo = getTodo(todos[0])

		userEvent.click(completedTodo)

		expect(completedTodo).toHaveStyle('color: #808080;')
		expect(completedTodo).toHaveStyle('text-decoration: line-through;')
	})
})
