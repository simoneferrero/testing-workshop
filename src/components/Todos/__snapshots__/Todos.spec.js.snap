// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given <Todos /> should render correctly 1`] = `
<div
  class="styled__StyledWrapper-p9f9rb-0 dzSefC"
  data-testid="todos"
>
  <form
    class="sc-bdVaJa blTatB"
  >
    <input
      class="sc-bwzfXH flCDJS"
      placeholder="Add TODO"
      value=""
    />
    <button
      class="sc-htpNat cbrJpO"
      type="submit"
    >
      ADD
    </button>
  </form>
  <label
    class="styled__StyledLabel-p9f9rb-1 iwEyoC"
    for="showCompleted"
  >
    Show Completed:
     
    <input
      checked=""
      id="showCompleted"
      type="checkbox"
    />
  </label>
  <div
    class="styled__StyledPaper-p9f9rb-2 hlILHM"
    data-testid="paper"
  />
</div>
`;
